core/vm/evm.go

```diff
@@ -135,11 +138,13 @@ type EVM struct {
 	// available gas is calculated in gasCall* according to the 63/64 rule and later
 	// applied in opCall*.
 	callGasTemp uint64
+
+	precompileMgr modules.Manager
 }
 
 // NewEVM returns a new EVM. The returned EVM is not thread safe and should
 // only ever be used *once*.
-func NewEVM(blockCtx BlockContext, txCtx TxContext, statedb StateDB, chainConfig *params.ChainConfig, config Config) *EVM {
+func NewEVM(blockCtx BlockContext, txCtx TxContext, statedb StateDB, chainConfig *params.ChainConfig, config Config, precompileMgr modules.Manager) *EVM {
```

Comment:
I was thinking more of just passing a slice of enabled precompiles at the given block -- NewEVM only used at a single height, `BlockContext` containing the fixed block height, so I don't think we need to dynamically query which contracts are enabled.

Similar to how chainRules will contain static fields of flags whether or not certain things are enabled -- I think main downside compared to precompileManager is performance since it's not lazily fetched but I think realistically it'll be a slice with only few addresses. If there's something else I'm missing lmk!

```suggestion
func NewEVM(blockCtx BlockContext, txCtx TxContext, statedb StateDB, chainConfig *params.ChainConfig, config Config, enabledPrecompiles []common.Address) *EVM {
```

Reply:
I had 2 thoughts:
- 1st one is precompileManager gives us more flexibility, maybe we'll need to extend it at some point with additional methods: IsInitialized(), Version(), etc... don't have real example as of now
- 2nd one is performance (you already mentioned), let's say we want to get all registered and enabled precompiles, with `enabledPrecompiles []common.Address` we'll have to iterate over 2 arrays in nested cycles
```go
for _, p := registry.GetRegisteredPrecompiles() {
   if slices.Contains(evm.enabledContracts, p.addr) {
     precompiles = append(precompiles, p)
   }
}
```
but I think performance impact is negligible, and overall your solutions is easier and cleaner

Reply:
If we want to add more metadata for each precompile we can pass in a slice of structs instead later on too, that include version info at the given height

Reply:
fixed